File: rust/examples/basic/src/processor.rs
use anyhow::Result;
use aptos_processor_framework::{
    indexer_protos::transaction::v1::Transaction, ProcessingResult, ProcessorTrait,
};
use tracing::debug;

#[derive(Debug)]
pub struct ExampleProcessor {}

/// A processor that just prints the txn version.
#[async_trait::async_trait]
impl ProcessorTrait for ExampleProcessor {
    fn name(&self) -> &'static str {
        "ExampleProcessor"
    }

    async fn process_transactions(
        &self,
        transactions: Vec<Transaction>,
        start_version: u64,
        end_version: u64,
    ) -> Result<ProcessingResult> {
        for transaction in transactions {
            debug!(
                "{}: Processing transaction {}",
                self.name(),
                transaction.version
            );
        }
        Ok((start_version, end_version))
    }
}

File: typescript/sdk/src/processor.ts
import { aptos } from "@aptos-labs/aptos-protos";
import { DataSource } from "typeorm";

/**
 * The result of processing a chunk of transactions. This is lets the prcoessor tell
 * the worker what range of transactions it processed.
 */
export type ProcessingResult = {
  startVersion: bigint;
  endVersion: bigint;
};

/**
 * A processor is given a batch of transactions. It is expected to process the
 * transactions, write some derived data to storage if appropriate, and return the
 * range of transactions it processed.
 */
export abstract class TransactionsProcessor {
  /**
   * Name of the processor for status logging and tracking of the latest processed
   * version. Prefer camel_case.
   */
  abstract name(): string;

  /**
   * Process transactions. The function is given the start and end versions of the
   * given chunk of transactions. It is expected to process the transactions, write
   * to storage if appropriate, and return the range of transactions it processed.
   */
  abstract processTransactions({
    transactions,
    startVersion,
    endVersion,
    dataSource,
  }: {
    transactions: aptos.transaction.v1.Transaction[];
    startVersion: bigint;
    endVersion: bigint;
    dataSource: DataSource; // DB connection
  }): Promise<ProcessingResult>;
}
